plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,404 @@
+	"SlicerT",
+	QT_TRANSLATE_NOOP("PluginBrowser", "Basic Slicer"),
+	"Daniel Kauss Serna <daniel.kauss.serna@gmail.com>",
+	0x0100,
+	Plugin::Type::Instrument,
+	new PluginPixmapLoader("logo"),
+	nullptr,
+	nullptr,
+};
+} // end extern
+
+// ################################# SlicerT ####################################
+
+SlicerT::SlicerT(InstrumentTrack* instrumentTrack)
+	: Instrument(instrumentTrack, &slicert_plugin_descriptor)
+	, m_noteThreshold(0.6f, 0.0f, 2.0f, 0.01f, this, tr("Note threshold"))
+	, m_fadeOutFrames(10.0f, 0.0f, 100.0f, 0.1f, this, tr("FadeOut"))
+	, m_originalBPM(1, 1, 999, this, tr("Original bpm"))
+	, m_sliceSnap(this, tr("Slice snap"))
+	, m_enableSync(false, this, tr("BPM sync"))
+	, m_originalSample()
+	, m_parentTrack(instrumentTrack)
+{
+	m_sliceSnap.addItem("Off");
+	m_sliceSnap.addItem("1/1");
+	m_sliceSnap.addItem("1/2");
+	m_sliceSnap.addItem("1/4");
+	m_sliceSnap.addItem("1/8");
+	m_sliceSnap.addItem("1/16");
+	m_sliceSnap.addItem("1/32");
+	m_sliceSnap.setValue(0);
+}
+
+void SlicerT::playNote(NotePlayHandle* handle, sampleFrame* workingBuffer)
+{
+	if (m_originalSample.frames() <= 1) { return; }
+
+	int noteIndex = handle->key() - m_parentTrack->baseNote();
+	const fpp_t frames = handle->framesLeftForCurrentPeriod();
+	const f_cnt_t offset = handle->noteOffset();
+	const int bpm = Engine::getSong()->getTempo();
+	const float pitchRatio = 1 / std::exp2(m_parentTrack->pitchModel()->value() / 1200);
+
+	float speedRatio = static_cast<float>(m_originalBPM.value()) / bpm;
+	if (!m_enableSync.value()) { speedRatio = 1; }
+	speedRatio *= pitchRatio;
+	speedRatio *= Engine::audioEngine()->processingSampleRate() / static_cast<float>(m_originalSample.sampleRate());
+
+	float sliceStart, sliceEnd;
+	if (noteIndex == 0) // full sample at base note
+	{
+		sliceStart = 0;
+		sliceEnd = 1;
+	}
+	else if (noteIndex > 0 && noteIndex < m_slicePoints.size())
+	{
+		noteIndex -= 1;
+		sliceStart = m_slicePoints[noteIndex];
+		sliceEnd = m_slicePoints[noteIndex + 1];
+	}
+	else
+	{
+		emit isPlaying(-1, 0, 0);
+		return;
+	}
+
+	if (!handle->m_pluginData) { handle->m_pluginData = new PlaybackState(sliceStart); }
+
+	float noteDone = ((PlaybackState*)handle->m_pluginData)->noteDone();
+	float noteLeft = sliceEnd - noteDone;
+
+	if (noteLeft > 0)
+	{
+		int noteFrame = noteDone * m_originalSample.frames();
+
+		SRC_STATE* resampleState = static_cast<PlaybackState*>(handle->m_pluginData)->resamplingState();
+		SRC_DATA resampleData;
+		resampleData.data_in = (m_originalSample.data() + noteFrame)->data();
+		resampleData.data_out = (workingBuffer + offset)->data();
+		resampleData.input_frames = noteLeft * m_originalSample.frames();
+		resampleData.output_frames = frames;
+		resampleData.src_ratio = speedRatio;
+
+		src_process(resampleState, &resampleData);
+
+		float nextNoteDone = noteDone + frames * (1.0f / speedRatio) / m_originalSample.frames();
+		static_cast<PlaybackState*>(handle->m_pluginData)->setNoteDone(nextNoteDone);
```

Comment:
```suggestion
	if (!handle->m_pluginData) { handle->m_pluginData = new PlaybackState(sliceStart); }
	auto playbackState = static_cast<PlaybackState*>(handle->m_pluginData);

	float noteDone = playbackState->noteDone();
	float noteLeft = sliceEnd - noteDone;

	if (noteLeft > 0)
	{
		int noteFrame = noteDone * m_originalSample.frames();

		SRC_STATE* resampleState = playbackState->resamplingState();
		SRC_DATA resampleData;
		resampleData.data_in = (m_originalSample.data() + noteFrame)->data();
		resampleData.data_out = (workingBuffer + offset)->data();
		resampleData.input_frames = noteLeft * m_originalSample.frames();
		resampleData.output_frames = frames;
		resampleData.src_ratio = speedRatio;

		src_process(resampleState, &resampleData);

		float nextNoteDone = noteDone + frames * (1.0f / speedRatio) / m_originalSample.frames();
		playbackState->setNoteDone(nextNoteDone);
```
Simplification - use `playbackState` variable instead of using `static_cast<PlaybackState*>(handle->m_pluginData)` every time it's needed.